session.go

```diff
@@ -700,6 +706,30 @@ func (s *Session) deallocateHandle(l *link) {
 	s.handles.Remove(l.handle)
 }
 
+func (s *Session) abandonLink(l *link) {
+	s.abandonedLinksMu.Lock()
```

Comment:
Why don't we just send out the detach frame now?

Reply:
I was experimenting with such a change shortly after posting this PR.  Will update when I have more info.

Reply:
Making this change runs into another bug I noticed yesterday and was planning on fixing post this PR.  I will make it now so that we don't have to delay sending of the detach frame.

Reply:
I've thought about this more and I think we should perform the cleanup during session/link creation.  The reason being, in order for the cleanup to be reliable/deterministic, we need to ensure that the handshake of frames actually happens; this requires a context and the ability to communicate the result to the caller in the event of failure.  I've built the "reliable writes" feature with this pattern and it's working quite well.

Reply:
Great point about making sure we have a spot to actually return the error.